spec/services/programmes/progress_query_spec.rb

```diff
@@ -0,0 +1,241 @@
+  let!(:other_activity) {
+    act = create(:activity)
+    create(:programme_activity, activity: act, programme_activity_grouping: other_course_group, programme: other_programme)
+    act
+  }
+  let!(:other_community_group_1) { create(:programme_activity_grouping, programme: other_programme, community: true) }
+  let!(:other_community_activity_1) {
+    act = create(:activity)
+    create(:programme_activity, activity: act, programme_activity_grouping: other_community_group_1, programme: other_programme)
+    act
+  }
+
+  def unenrolled_user
+    user = create(:user)
+    create(:user_programme_enrolment, user:, programme: other_programme)
+    user
+  end
+
+  def enrolled_user
+    user = create(:user)
+    create(:user_programme_enrolment, user:, programme:)
+    user
+  end
+
+  def active_achievement(user, activity)
+    date = Faker::Time.between(from: DateTime.now.months_ago(4), to: DateTime.now.days_ago(1))
+    create(:completed_achievement, activity:, user:, created_at: date, updated_at: date)
+    user
+  end
+
+  def lapsing_achievement(user, activity)
+    date = Faker::Time.between(from: DateTime.now.months_ago(8), to: DateTime.now.months_ago(4))
+    create(:completed_achievement, activity:, user:, created_at: date, updated_at: date)
+    user
+  end
+
+  def lapsed_achievement(user, activity)
+    date = Faker::Time.between(from: DateTime.now.months_ago(12), to: DateTime.now.months_ago(8))
+    create(:completed_achievement, activity:, user:, created_at: date, updated_at: date)
+    user
+  end
+
+  it "enrolled_user should not have user_programme_enrolment" do
+    # Test is to ensure that nothing has changed with auto enrolment that might impact this test suite
+    expect(UserProgrammeEnrolment.where(user: unenrolled_user, programme:)).to be_empty
+  end
+
+  context "dates_by_state" do
+    it "should return correct dates for active" do
+      active_range = described_class.new(programme, :active, true).dates_by_state
+      expect(active_range.first.to_date).to eq(DateTime.now.months_ago(4).to_date)
+      expect(active_range.last.to_date).to eq(DateTime.now.to_date)
+    end
+
+    it "should return correct dates for lasping" do
+      lapsing_range = described_class.new(programme, :lapsing, true).dates_by_state
+      expect(lapsing_range.first.to_date).to eq(DateTime.now.months_ago(8).to_date)
+      expect(lapsing_range.last.to_date).to eq(DateTime.now.months_ago(4).to_date)
+    end
+    it "should return correct dates for lasping" do
+      lapsed_range = described_class.new(programme, :lapsed, true).dates_by_state
+      expect(lapsed_range.first.to_date).to eq(DateTime.now.months_ago(1000).to_date)
+      expect(lapsed_range.last.to_date).to eq(DateTime.now.months_ago(8).to_date)
+    end
+
+    it "should default to active range for non matching key" do
+      active_range = described_class.new(programme, :non_sense, true).dates_by_state
```

Comment:
Would it be clearer to call the non matching key "unknown_key" or something similar?